win32ss/user/user32/windows/cursoricon.c

```diff
@@ -267,57 +267,34 @@ CURSORICON_ConvertPngToBmpIcon(
-       NtUserSetSystemCursor(LoadImageW( 0, IDC_NO,          IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE ), OCR_NO);
-       NtUserSetSystemCursor(LoadImageW( 0, IDC_HAND,        IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE ), OCR_HAND);
-       NtUserSetSystemCursor(LoadImageW( 0, IDC_APPSTARTING, IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE ), OCR_APPSTARTING);
-       NtUserSetSystemCursor(LoadImageW( 0, IDC_HELP,        IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE ), OCR_HELP);
-   }
-}
-
 /* This callback routine is called directly after switching to gui mode */
 NTSTATUS
 WINAPI
 User32SetupDefaultCursors(PVOID Arguments,
                           ULONG ArgumentLength)
 {
-    BOOL *DefaultCursor = (BOOL*)Arguments;
-    HCURSOR hCursor;
-
-    /* Load system cursors first */
-    LoadSystemCursors();
-
-    if(*DefaultCursor)
-    {
-        /* set default cursor */
-        hCursor = LoadCursorW(0, IDC_ARROW);
-        SetCursor(hCursor);
-    }
-    else
-    {
-        /* FIXME load system cursor scheme */
-        SetCursor(0);
-        hCursor = LoadCursorW(0, IDC_ARROW);
-        SetCursor(hCursor);
-    }
-
-    return(ZwCallbackReturn(&hCursor, sizeof(HCURSOR), STATUS_SUCCESS));
+    PLOADCURSORS_CALLBACK_ARGUMENTS Common = Arguments;
+
+    ERR("Loading System Cursors\n");
+
+    Common->hCursorArrow       = LoadImageW(NULL, IDC_ARROW,       IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorIbeam       = LoadImageW(NULL, IDC_IBEAM,       IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorWait        = LoadImageW(NULL, IDC_WAIT,        IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorCross       = LoadImageW(NULL, IDC_CROSS,       IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorUp          = LoadImageW(NULL, IDC_UPARROW,     IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorIcon        = LoadImageW(NULL, IDC_ICON,        IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorSize        = LoadImageW(NULL, IDC_SIZE,        IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorSizeNwse    = LoadImageW(NULL, IDC_SIZENWSE,    IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorSizeNesw    = LoadImageW(NULL, IDC_SIZENESW,    IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorSizeWe      = LoadImageW(NULL, IDC_SIZEWE,      IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorSizeNs      = LoadImageW(NULL, IDC_SIZENS,      IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorSizeAll     = LoadImageW(NULL, IDC_SIZEALL,     IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorNo          = LoadImageW(NULL, IDC_NO,          IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorHand        = LoadImageW(NULL, IDC_HAND,        IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorAppStarting = LoadImageW(NULL, IDC_APPSTARTING, IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
+    Common->hCursorHelp        = LoadImageW(NULL, IDC_HELP,        IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE);
```

Comment:
On Windows, it appears that these (the loading of cursors) are indeed done from user32; however, not as a callback as done here, but instead in a direct manner:
- when the user-mode client part of the win32 subsystem connects to win32k (for us, this is in `win32ss/user/user32/misc/dllmain.c!ClientThreadSetupHelper` in the `if (gfServerProcess && IsFirstThread)` block);
- in CreateWindowStationW in case a non-empty window station name is provided _**AND**_ `gfServerProcess` is TRUE.

Following these cursors loading, a call to `NtUserCallNoParam(NOPARAM_ROUTINE_LOADCURSANDICOS);` is then made, in order to synchronize in win32k these cursors loaded from user32.